### PR TITLE
fix(prometheus source): rewrite parser, improve error handlings 

### LINKED ISSUE #2668: How to debug prometheus source
```
~ # RUST_LOG=debug vector 
May 22 13:02:16.920  INFO vector: Log level "info" is enabled.
May 22 13:02:16.920  INFO vector: Loading configs. path=["/etc/vector/vector.toml"]
May 22 13:02:16.923  INFO vector: Vector is starting. version="0.9.1" git_version="v0.9.1" released="Thu, 30 Apr 2020 08:24:03 +0000" arch="x86_64"
May 22 13:02:16.923  INFO vector::topology: Running healthchecks.
May 22 13:02:16.923  INFO vector::topology: Starting source "prometheus"
May 22 13:02:16.923  INFO vector::topology::builder: Healthcheck: Passed.
May 22 13:02:16.923  INFO vector::topology: Starting sink "stdout"
May 22 13:02:16.937 ERROR source{name=prometheus type=prometheus}: vector::internal_events::prometheus: parsing error error=Malformed
^CMay 22 13:02:18.401  INFO vector: Shutting down.
```

There are tools/commands to print what is malformed to debug this?

### LINKED ISSUE #3160: Prometheus source minor tech debt
While reviewing #3090, I saw a few pieces of poor coding in the parser:

1. `enum ParserError` doesn't actually enumerate all the error cases. Instead, it uses a string embedded in the code to describe how something was malformed. Since there are a fixed set of ways something can be malformed, they should be enumerated explicitly in the error type.
2. Quoted values have leading and trailing spaces trimmed. Quoted values shouldn't be trimmed at all.
3. Values with a single leading or trailing double quote will have that stripped.

Seeing these, I'm sure there are other gotchas to be cleaned up. This could use a fuller review for additional issues.

### LINKED ISSUE #3276: The `prometheus` source fails to parse valid prometheus metrics
I was helping a customer debug the following error for the `prometheus` source:

```
Jul 30 19:51:27.634 ERROR source{name=mystique type=prometheus}: vector::internal_events::prometheus: parsing error error=Malformed
```

He sent the following response from the configured endpoint that Vector was scraping:

```
curl -v -XGET http://localhost:9913/metrics
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9913 (#0)
> GET /metrics HTTP/1.1
> Host: localhost:9913
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: text/plain; version=0.0.4; charset=utf-8
< Date: Thu, 30 Jul 2020 20:00:30 GMT
< Transfer-Encoding: chunked
< 
# HELP nginx_server_bytes request/response bytes
# TYPE nginx_server_bytes counter
nginx_server_bytes{direction="in",host="*"} 263719
nginx_server_bytes{direction="in",host="_"} 255061
nginx_server_bytes{direction="in",host="nginx-vts-status"} 8658
nginx_server_bytes{direction="out",host="*"} 944199
nginx_server_bytes{direction="out",host="_"} 360775
nginx_server_bytes{direction="out",host="nginx-vts-status"} 583424
# HELP nginx_server_cache cache counter
# TYPE nginx_server_cache counter
nginx_server_cache{host="*",status="bypass"} 0
nginx_server_cache{host="*",status="expired"} 0
nginx_server_cache{host="*",status="hit"} 0
nginx_server_cache{host="*",status="miss"} 0
nginx_server_cache{host="*",status="revalidated"} 0
nginx_server_cache{host="*",status="scarce"} 0
```

Vector should be able to parse this data. I'd like to add a test specifically for this data set.

### PR DESCRIPTION
**Pull request by juchiast**
closes #3276 
closes #3160
closes #2668 

- Rewrite parser using nom.
- Print malformed response in `debug!`.
- Treat non-200 response as error.
- Add `url` to error events, for better debuggability  when use uses multiple hosts.

Signed-off-by: Duy Do <juchiast@gmail.com>

### CODE REVIEW

File: src/sources/prometheus/parser.rs
Comment (juchiast) [inline src/sources/prometheus/parser.rs:1217]: @binarylogic The test case you mentioned, it doesn't fail though.

**Comment by juchiast** [comment]
Format reference: https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md

File: src/sources/prometheus/parser.rs
Comment (binarylogic) [reply src/sources/prometheus/parser.rs:1279]: Hm, that's strange. I'd be curious to see if it fails during an actual integration test with the `prometheus` source. Seeing that we don't have an integraiton test for this source I've opened #3289. What do you think?

Comment (juchiast) [reply src/sources/prometheus/parser.rs:1279]: It would be great if we can get more details about the error (via #3173).

File: src/sources/prometheus/parser.rs
Comment (juchiast) [inline src/sources/prometheus/parser.rs:551]: We don't support this anymore.

File: src/sources/prometheus/parser.rs
Comment (juchiast) [inline src/sources/prometheus/parser.rs:127]: Do all lines for a given tags have to be consecutive? If yes, we can get rig of `IndexMap` and just use a `Vec` (like in `fn group_metrics`).

**Review by juchiast — COMMENTED** [review]
I moved `fn group_metrics` out and rewrite it. All test cases passed.

**Comment by binarylogic** [comment]
@juchiast nice work so far. How are the error messages when parsing fails? It’s be nice to see some examples and possibly test for them since it helps users debug malformed endpoints when scraping.

**Comment by juchiast** [comment]
@binarylogic I improved error handling and added some tests, printing errors in `prometheus-parser`' tests.

**Comment by juchiast** [comment]
- [ ] Add test for `Transfer-Encoding: chunked`
- [ ] Add test for `gzip`.
- [ ] Write a streaming parser. (Is it worth it?)

File: lib/prometheus-parser/src/line.rs
Comment (bruceg) [inline lib/prometheus-parser/src/line.rs:357]: Ditto source error?

File: lib/prometheus-parser/src/line.rs
Comment (bruceg) [inline lib/prometheus-parser/src/line.rs:612]: Similar comments about asserting specific error matches here.

File: src/sources/prometheus/parser.rs
Comment (bruceg) [inline src/sources/prometheus/parser.rs:138]: Does this still need checking?

**Review by bruceg — CHANGES_REQUESTED** [review]
I'm glad to see this expanded and moved into a lib. Some things need work here though.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs:21]: Internal bugs, they should be assertion but we don't seem to use assert in our code?

File: lib/prometheus-parser/src/line.rs
Comment (juchiast) [reply lib/prometheus-parser/src/line.rs:274]: Nom errors are single enum variants (https://docs.rs/nom/5.1.2/nom/error/enum.ErrorKind.html), it is not very meaningful and I don't think we should include it. In this case it is an `ErrorKind::Takewhile1`.

File: lib/prometheus-parser/src/line.rs
Comment (juchiast) [reply lib/prometheus-parser/src/line.rs:284]: Same here, an `ErrorKind::Char`.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs:19]: Nom drops the original error so we don't have access to that.

File: lib/prometheus-parser/src/line.rs
Comment (juchiast) [reply lib/prometheus-parser/src/line.rs:198]: I tried `separated_list` but it has several  problems, such as it treats `parse_name_value`'s error as the list has ended, and drop that error, there are more but I don't remember the problems.

File: lib/prometheus-parser/src/line.rs
Comment (juchiast) [reply lib/prometheus-parser/src/line.rs:37]: `ErrorKind` doesn't implement `Error`, so we can't use `#[snafu(source)]`.

File: src/sources/prometheus/parser.rs
Comment (juchiast) [reply src/sources/prometheus/parser.rs:138]: @bruceg I'm not sure, negative values here will be rounded to 0. Should we report an error for this?

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: I moved `labels` out.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: The error should be `ErrorKind::ParseFloatError`, I changed it to use `with_context` just in case it is not.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: done.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: done.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: done.

File: lib/prometheus-parser/src/line.rs
Comment (juchiast) [reply lib/prometheus-parser/src/line.rs]: done.

File: lib/prometheus-parser/src/line.rs
Comment (juchiast) [reply lib/prometheus-parser/src/line.rs]: Removed this variant, now `nom::Err::Failure` is treated the same as `nom::Err::Error`.

The code should not emit `nom::Err::Failure` at the moment, but this is implementation-specific.

File: lib/prometheus-parser/src/lib.rs
Comment (bruceg) [reply lib/prometheus-parser/src/lib.rs]: We don't use `assert!` directly except in tests. However, there are some instances of `.expect` which will panic, and `unreachable!`, which better express that idea. If there are conditions in the parser that should be impossible, use one of those and add a note about why it is impossible. Things that shouldn't happen in practice but could be triggered by user or attacker controlled data should still be covered by an error.

File: lib/prometheus-parser/src/line.rs
Comment (bruceg) [reply lib/prometheus-parser/src/line.rs:271]: Ok, got it. FTR @Jeffail went with [pest](https://crates.io/crates/pest) in part because of the error message story. No need for you to convert though, it would probably be overkill here.

File: src/sources/prometheus/parser.rs
Comment (bruceg) [reply src/sources/prometheus/parser.rs]: Probably, yes. Should at least output a warning.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: Removed `InvalidName`, `Incomplete` repaced with `unreachable!()`.

File: src/sources/prometheus/parser.rs
Comment (juchiast) [reply src/sources/prometheus/parser.rs]: I added `ExpectNonNegativeValue`.

**Comment by juchiast** [comment]
Hi @bruceg @binarylogic how's this PR?

**Comment by binarylogic** [comment]
Thanks @juchiast. The error messages look good to me based on the tests. I'd like for @bruceg to give the signoff on the code before merging. Once he approves feel free to self merge.

File: src/sources/prometheus/parser.rs
Comment (bruceg) [inline src/sources/prometheus/parser.rs:22]: The logic of this function is strongly tied into that of the rest of the parsing that was moved into a separate workspace. Was there a reason to keep this one here?

File: src/sources/prometheus/parser.rs
Comment (bruceg) [inline src/sources/prometheus/parser.rs:38]: This pattern is repeated several times here and could use a helper to clean that up.

**Review by bruceg — CHANGES_REQUESTED** [review]
A few smaller issues here, not counting the possibility of moving all of `parser.rs` into the new lib. Getting better.

**Comment by bruceg** [comment]
In the future, please avoid doing force pushes to a branch that is up for re-review. It makes it harder for reviewers to see what has changed since their last review, particularly on a PR like this with large code changes. We use squash merges (and tidy up the resulting commit message) so amending past commits makes little difference to the end result.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs:99]: We should also handle this case, but I think it's better to use `u64` in `MetricValue`.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs:129]: Let's keep it like this.

**Comment by binarylogic** [comment]
@juchiast a couple of more questions regarding HTTP status codes:

1. How are non 2XX responses handled? I've worked with customers that encounter 504s, for example, and it was not clear to them until they hit the endpoint directly.
2. How about redirects?

File: src/sources/prometheus/mod.rs
Comment (binarylogic) [reply src/sources/prometheus/mod.rs:118]: I agree with @bruceg, is there any way we could include the `packet` as an attribute on the error and then move this `debug` statement [here](https://github.com/timberio/vector/pull/3286/files#diff-de68ffa2653a73a605408a190e2cc024R62)? Are there performance concerns with this?

Comment (binarylogic) [reply src/sources/prometheus/mod.rs:118]: Actually, I'll defer to @bruceg on this. I can also see the case for including an inline `debug` statement for this purpose.

**Comment by juchiast** [comment]
@binarylogic anything other than 200 is currently considered as error, including redirects (3XX codes). The error message is `PrometheusErrorResponse`:

```rust
error!(message = "HTTP error response.", url = %self.url, code = %self.code);
```

**Comment by binarylogic** [comment]
@juchiast works for me 👍

**Review by binarylogic — APPROVED** [review]
Error messages look great to me. Once @bruceg approves of the final changes we can merge.

File: src/sources/prometheus/mod.rs
Comment (bruceg) [reply src/sources/prometheus/mod.rs:118]: I suggest adding a reference to the packet to the event structure, and then move the debug into `fn emit_logs`.

File: src/sources/prometheus/parser.rs
Comment (bruceg) [reply src/sources/prometheus/parser.rs:31]: Right, I've been working on separating out the whole event tree, but until that point this part has to stay here.

File: lib/prometheus-parser/src/lib.rs
Comment (bruceg) [reply lib/prometheus-parser/src/lib.rs:99]: That may be true, but has implications beyond this issue. Handle this as a `u32` here which will need the overflow check, and then open an issue to discuss the `u64` proposal.

File: src/sources/prometheus/mod.rs
Comment (juchiast) [reply src/sources/prometheus/mod.rs]: Moved.

File: lib/prometheus-parser/src/lib.rs
Comment (juchiast) [reply lib/prometheus-parser/src/lib.rs]: I added `ParserError::ValueOutOfRange`.

File: src/internal_events/prometheus.rs
Comment (bruceg) [inline src/internal_events/prometheus.rs:59]: Even though the source `body` is this type (through `from_utf8_lossy`), you could probably have gotten away with `body: &'a str` here, since `Cow<'a, str>` dereferences to that type. This works, though.

**Review by bruceg — APPROVED** [review]

**Comment by bruceg** [comment]
Oh, there's a merge conflict left to resolve. The price of change, I guess.

**Pull request merged**
